### PR TITLE
[FLINK-3498] Implement TRIM, SUBSTRING as reference design for Table API

### PR DESCRIPTION
**Pull request by twalthr**
This PR implements `TRIM` and `SUBSTRING` for the Table API. It generalizes all calls to the expression `Call` which allows to add functions easier and override functions later. `TRIM` is an example for a special function as it uses some symbols like `BOTH`, `LEADING` etc. and special syntax. I extended the `ExpressionEvaluator`, it now also tests SQL syntax.

After this PR has been merged, I would continue with this design for FLINK-3497.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/scala/table/expressionDsl.scala [hunk 1]
```diff
@@ -63,17 +63,77 @@ trait ImplicitExpressionOperations {
...
-  def as(name: Symbol) = Naming(expr, name.name)
+  /**
+    * Removes leading and/or trailing characters from the given string.
+    *
+    * @param removeLeading if true, remove leading characters
+    * @param removeTrailing if true, remove trailing characters
+    * @param character String containing the character
+    * @return trimmed string
+    */
+  def trim(removeLeading: Boolean, removeTrailing: Boolean, character: Expression) = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/scala/table/expressionDsl.scala]: Can we remove the `trim()` method if we add the default values to the parameters in this function?

Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/scala/table/expressionDsl.scala]: Yes, makes sense. I updated the PR.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/calls/ScalarFunctions.scala [hunk 2]
```diff
@@ -49,6 +49,11 @@ object ScalarFunctions {
     STRING_TYPE_INFO,
     BuiltInMethod.SUBSTRING.method)
 
+  addSqlFunctionTrim(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/calls/ScalarFunctions.scala]: Can you briefly explain why `TRIM` cannot be handled as a generic SQL method like `SUBSTRING` and requires its own `TrimCallGenerator`?

**Comment by fhueske** [comment]
The changes look good. However, it still looks like quite a bit of code at various places to add a new function. For example, why do we need a n own code generator for `TRIM`? I thought we could leverage Calcite's build-in functions for that or is there none for `TRIM`?

**Comment by twalthr** [comment]
"quite a bit of code at various places": Yes, you are right. I already tried to simplify it as much as I can. The problem is that we have 3 APIs that all have slightly different requirements. I chose `TRIM` because it is an extreme example. Other functions (80%) are easier, because they don't need their own code generator and ExpressionParser.

My current implementation has a proper separation of the different layers:
Expression API -> RexNodeTranslator -> CodeGen

3 layers + 3 APIs = up to 6 files to change for every function.

I don't know if it makes sense to mix the layers.

**Comment by fhueske** [comment]
OK, I see. Sounds good to me.
+1 to merge.

**Comment by fhueske** [comment]
Will merge this to `tableOnCalcite` branch.

**Comment by fhueske** [comment]
Merged to `tableOnCalcite`. Can you close the PR @twalthr?

**Pull request closed without merging**